### PR TITLE
DAOS-8722 dfuse: Improve dfuse readdir code.

### PR DESCRIPTION
**Pull request by ashleypittman**
Extend readdir testing to include multiple streams open concurrently.
Do not duplicate inode handle for open directories.
Create a seperate readdir handle rather than having it as part of the
directory handle.

This paves the way for sharing readdir handles across streams in the future.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

### CODE REVIEW

**Review by jolivier23 — APPROVED** [review]

File: utils/node_local_test.py [hunk 1]
```diff
@@ -1897,6 +1893,104 @@ def readdir_test(self, count, test_all=False):
...
+        assert len(files) == count
+
+    @needs_dfuse
+    def test_readdir_unlink(self):
+        """Test readdir where a entry is removed mid read
+
+        Populate a directory, read the contents to know the order, then unlink a file and re-read
+        to verify the file is missing.  If doing the unlink during read then the kernel cache
+        will include the unlinked file so do not check for this behavior.
+        """
```
Comment (daltonbohning) [inline utils/node_local_test.py:1969]: Is this test actually unlinking **_during_** read?

**Review by daltonbohning — COMMENTED** [review]
Should this run with `Features: dfuse,-manual` to  prevent accidental regression?

File: utils/node_local_test.py (hunk 1 above)
Comment (ashleypittman) [reply utils/node_local_test.py:1969]: during readdir, yes.  Whilst the directory is open.

Comment (daltonbohning) [reply utils/node_local_test.py:1969]: Ah, "mid read" means "mid readdir" :)

**Review by daltonbohning — APPROVED** [review]

**Pull request merged**